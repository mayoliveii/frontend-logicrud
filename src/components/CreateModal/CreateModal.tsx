/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  Button,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalCloseButton,
  ModalBody,
  ModalFooter,
  Box,
  FormControl,
  FormLabel,
  Input,
  InputGroup,
  InputLeftElement,
  Textarea,
  useColorModeValue,
  VStack,
} from '@chakra-ui/react';
import { FaTruck } from 'react-icons/fa';
import { useState } from 'react';
import apiService from '../../apiServices/apiService';

interface CreateModalProps {
  isOpen: boolean;
  onClose: () => void;
  onCategoryAdded?: () => void;
}

export function CreateModal({ isOpen, onClose, onCategoryAdded }: CreateModalProps) {
  const [formState, setFormState] = useState({ title: '', description: '' });

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setFormState({
      ...formState,
      [event.target.name]: event.target.value,
    });
  };

  const handleSubmit = async () => {
    try {
      const newCategory = await apiService.createCategory(formState);

      setFormState({ title: '', description: '' });

      onClose();
      if (onCategoryAdded) {
        onCategoryAdded();
      }

      return newCategory;
    } catch (error: any) {
      throw new Error(`Error creating category: ${error}`);
    }
  };

  const renderNameInput = () => (
    <FormControl isRequired>
      <FormLabel>Nome</FormLabel>
      <InputGroup>
        <InputLeftElement>
          <FaTruck />
        </InputLeftElement>
        <Input
          type="text"
          name="title"
          placeholder="Your Name"
          value={formState.title}
          onChange={handleInputChange}
        />
      </InputGroup>
    </FormControl>
  );

  const renderDescriptionInput = () => (
    <FormControl isRequired>
      <FormLabel>Descrição</FormLabel>
      <Textarea
        name="description"
        placeholder="Your Message"
        rows={6}
        resize="none"
        value={formState.description}
        onChange={handleInputChange}
      />
    </FormControl>
  );

  return (
    <Modal isOpen={isOpen} onClose={onClose}>
      <ModalOverlay />
      <ModalContent>
        <ModalHeader>Adicionar nova categoria de veículo</ModalHeader>
        <ModalCloseButton />
        <ModalBody>
          <Box
            bg={useColorModeValue('white', 'gray.700')}
            borderRadius="lg"
            p={8}
            color={useColorModeValue('gray.700', 'whiteAlpha.900')}
            shadow="base"
          >
            <VStack spacing={5}>
              {renderNameInput()}
              {renderDescriptionInput()}
            </VStack>
          </Box>
        </ModalBody>
        <ModalFooter>
          <Button bg="#FF7A50" mr={3} color={'white'} _hover={{ bg: '#FF5823' }} onClick={handleSubmit}>
            Salvar
          </Button>
          <Button variant="ghost" onClick={onClose}>
            cancelar
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
}