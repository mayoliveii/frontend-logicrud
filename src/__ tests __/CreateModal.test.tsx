import { render, fireEvent, waitFor } from '@testing-library/react';
import { ChakraProvider } from '@chakra-ui/react';
import '@testing-library/jest-dom';
import { CreateModal } from '../components/CreateModal/CreateModal';

jest.mock('./../apiServices/apiService', () => ({
  createCategory: jest.fn(),
}));

const onClose = jest.fn();
const onCategoryAdded = jest.fn();

describe('CreateModal component', () => {
  it('renders CreateModal component and matches snapshot', () => {
    const { asFragment } = render(
      <ChakraProvider>
        <CreateModal isOpen={true} onClose={jest.fn()} />
      </ChakraProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles input and form submission', async () => {
    const { getByPlaceholderText, getByText } = render(
      <ChakraProvider>
        <CreateModal isOpen={true} onClose={onClose} onCategoryAdded={onCategoryAdded} />
      </ChakraProvider>
    );

    fireEvent.change(getByPlaceholderText('Your Name'), { target: { value: 'Test' } });
    fireEvent.change(getByPlaceholderText('Your Message'), { target: { value: 'Test' } });
    fireEvent.click(getByText('Salvar'));

    await waitFor(() => expect(onClose).toHaveBeenCalled());
    expect(onCategoryAdded).toHaveBeenCalled();
  });

  it('handles form submission error', async () => {
    const { getByPlaceholderText, getByText } = render(
      <ChakraProvider>
        <CreateModal isOpen={true} onClose={onClose} onCategoryAdded={onCategoryAdded} />
      </ChakraProvider>
    );

    fireEvent.change(getByPlaceholderText('Your Name'), { target: { value: 'Test' } });
    fireEvent.change(getByPlaceholderText('Your Message'), { target: { value: 'Test' } });
    fireEvent.click(getByText('Salvar'));

    await waitFor(() => expect(onClose).toHaveBeenCalled());
    expect(onCategoryAdded).toHaveBeenCalled();
  });
});