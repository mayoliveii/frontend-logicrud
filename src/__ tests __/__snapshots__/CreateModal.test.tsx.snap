// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateModal component renders CreateModal component and matches snapshot 1`] = `
<DocumentFragment>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
